### PR TITLE
Optimize /proc traversal

### PR DESCRIPTION
**Pull request by inercia**
The objective of this PR is to reduce the number of syscalls when collecting info from `/proc`. This PR does that by: 

* reading the required files (`/proc/*/[cmdline, comm, stat, fd/*]`, `/proc/net/tcp*`) once per cycle (`spy.interval`), parsing and caching the processes/connections information
* keeping a cache of open files for `/proc/*/[cmdline, comm, stat]` so we do not `open()`/`close()` these files so often.
* removing unnecessary `stat()`s when reading `/proc` directories

Changes:

* the required code for reading/parsing `/proc` has been proved from `procspy` to `/probe/proc`
* the `Walker` has been replaced by a `proc.Reader` that provides both `Processes()` and `Connections()`.  Users can walk the list of processes/connections in the same way by providing a function to these functions. A `CachingReader` can be used for caching and getting this info, updated when `Update()`. The code for reading and parsing connection and processes has been simplified.
* a `CachingReader` is initialized in `/probe/main.go` and used in both `endpoint.Reporter` and `process.Reporter` for reporting about processed and connections, and updated periodically by calling `Update()` every `spy.interval` seconds.
* unit tests have been improved by mocking things like file system operations
* things like `Proc` and `Process` have been unified, and many things from `/probe/process` have been moved to `/probe/proc`

This is part of the solution to #284

### CODE REVIEW

**Comment by 2opremio** [comment]
@inercia I don't know if it provides all you need but maybe you could use (and even extend)  https://github.com/mitchellh/go-ps

**Comment by 2opremio** [comment]
Actually, I don't think `go-ps` provides connection/file information, but https://github.com/c9s/goprocinfo certainly does and it seems more mature.

**Comment by inercia** [comment]
@2opremio thanks with the hint... `goprocinfo` seems interesting. The only problem I see is that they use `Read()`s instead of our current mechanism where we `buf.ReadFrom(someFile)` and the `buf` is obtained from a `Pool`, so we do not generate too much garbage... it's worth some investigation, though...

File: probe/proc/reader.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+	Cmdline   string
+	Threads   int
+	Inodes    []uint64
+}
+
+// ProcReader is something that reads the /proc directory and
+// returns some info like processes and connections
+type ProcReader interface {
+	// Processes walks through the processes
+	Processes(func(Process)) error
```
Comment (inercia) [inline probe/proc/reader.go]: Maybe `WalkProcesses` and `WalkConnections`?

**Comment by inercia** [comment]
@tomwilkie or @peterbourgon could you please review this?

**Comment by peterbourgon** [comment]
@inercia Sure. Would you mind giving a short overview of what this PR does, including motivation & pointers to key types/functions/changes?

File: probe/proc/reader_linux.go [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+	}
+
+	return nil
+}
+
+var bufPool = sync.Pool{
+	New: func() interface{} {
+		return bytes.NewBuffer(make([]byte, 0, 5000))
+	},
+}
```
Comment (inercia) [inline probe/proc/reader_linux.go]: Not sure we should keep this `Pool` anymore: we do not read the `/proc` concurrently from multiple goroutines (we only do it when `Update()`ing the `CachingReader`), so we could use only one `bytes.Buffer` and get rid of all the `Pool.Get`/`Pool.Put`/`Buffer.Reset` cycle...

**Comment by inercia** [comment]
@peterbourgon I have updated the PR description, so please let me know what you think about this...

**Comment by inercia** [comment]
This is the latest profile graph from this branch (using go 1.5). Comapred with the original graph in #284, we can see that the number of `syscalls` has been vastly reduced, but we sped a lot of time in `runtime.*` now (too much garbage, maybe?)...

profile

**Comment by peterbourgon** [comment]
@inercia Thanks, that helps a lot. I'll give it a look.

File: probe/proc/net.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+func (c Connection) Copy() Connection {
+	dupIP := func(ip net.IP) net.IP {
+		dup := make(net.IP, len(ip))
+		copy(dup, ip)
+		return dup
+	}
+
+	c.LocalAddress = dupIP(c.LocalAddress)
+	c.RemoteAddress = dupIP(c.RemoteAddress)
+	return c
```
Comment (peterbourgon) [inline probe/proc/net.go]: What about the Process.Inodes?

File: probe/proc/reader_linux.go (hunk 2 above)
Comment (peterbourgon) [inline probe/proc/reader_linux.go]: Yes, putting the buffer into the reader struct? If so, +1.

File: probe/proc/reader_linux.go [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+
+// Connections walks through all the connections in the "/proc"
+func (w *reader) Connections(withProcs bool, f func(Connection)) error {
```
Comment (peterbourgon) [inline probe/proc/reader_linux.go]: Presumably `r *reader` is meant.

File: probe/proc/reader_linux.go [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+		return err
+	}
+
+	buf := bytes.NewBuffer(make([]byte, 0, 5000))
+
+	for _, subdir := range dirEntries {
+		readIntoBuffer := func(filename string) error {
+			buf.Reset()
+			res := w.handles.ReadInto(path.Join(subdir, filename), buf)
+			return res
```
Comment (peterbourgon) [inline probe/proc/reader_linux.go]: `return w.handles.ReadInto(...)`

File: probe/proc/reader_linux.go [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+// instances of Process, which it then passes one-by-one to the
+// supplied function. Processes() is only made public so that is
+// can be tested.
+func (w *reader) Processes(f func(Process)) error {
+	dirEntries, err := w.proc.ReadDirNames(w.proc.Root())
+	if err != nil {
+		return err
+	}
+
+	buf := bytes.NewBuffer(make([]byte, 0, 5000))
```
Comment (peterbourgon) [inline probe/proc/reader_linux.go]: Why the explicit initialization?

File: probe/proc/reader_linux.go [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+
+		fdBase := path.Join(w.proc.Root(), strconv.Itoa(pid), "fd")
+		fdNames, err := w.proc.ReadDirNames(fdBase)
+		if err != nil {
+			return err
+		}
+
+		inodes := []uint64{}
+		for _, fdName := range fdNames {
+			var fdStat syscall.Stat_t
```
Comment (peterbourgon) [inline probe/proc/reader_linux.go]: Can this be safely moved out of the loop and re-used, perhaps?

File: probe/proc/reader_linux.go [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+	filesCacheExpiration = 60 * time.Second
+)
+
+var tcpFiles = []string{
+	"net/tcp",
+	"net/tcp6",
+}
+
+// A cache for files handles
+// Note: not intended to be used from multiple goroutines
```
Comment (peterbourgon) [inline probe/proc/reader_linux.go]: Wouldn't it be better to make it goroutine safe, than put this warning?

File: probe/proc/reader_linux.go [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/bluele/gcache"
+)
+
+const (
+	filesCacheLen        = 512
+	filesCacheExpiration = 60 * time.Second
```
Comment (peterbourgon) [inline probe/proc/reader_linux.go]: Just to be clear, a single file is for example /proc/<pid>/something, is that correct? If so, is 60 seconds perhaps too long?

File: probe/proc/reader_darwin.go [hunk 10]
```diff
@@ -44,6 +44,59 @@ func (walker) Walk(f func(Process)) error {
 	return nil
 }
 
+func (w *reader) Connections(withProcs bool, f func(Connection)) error {
```
Comment (peterbourgon) [inline probe/proc/reader_darwin.go]: Presumably `r *reader` is meant here (further comments like this elided)

File: probe/main.go [hunk 11]
```diff
@@ -186,7 +186,7 @@ func main() {
 				r = report.MakeReport()
 
 			case <-spyTick:
-				if err := processCache.Update(); err != nil {
+				if err := procReader.Update(*spyProcs); err != nil {
```
Comment (peterbourgon) [inline probe/main.go]: Great. This will help.

**Comment by peterbourgon** [comment]
The shape of it is good. Back to @inercia for those few comments (and rebasing, I guess).

File: probe/proc/reader_linux.go (hunk 2 above)
Comment (inercia) [inline probe/proc/reader_linux.go]: Done.

File: probe/proc/net.go (hunk 3 above)
Comment (inercia) [inline probe/proc/net.go]: Fixed. Added a `Process.Copy()`

File: probe/proc/reader_linux.go (hunk 4 above)
Comment (inercia) [inline probe/proc/reader_linux.go]: Right. Fixed.

File: probe/proc/reader_linux.go (hunk 5 above)
Comment (inercia) [inline probe/proc/reader_linux.go]: Fixed.

File: probe/proc/reader_linux.go (hunk 6 above)
Comment (inercia) [inline probe/proc/reader_linux.go]: Because I copied&pasted... :smile: Fixed.

File: probe/proc/reader_linux.go (hunk 7 above)
Comment (inercia) [inline probe/proc/reader_linux.go]: Done.

File: probe/proc/reader_linux.go (hunk 8 above)
Comment (inercia) [inline probe/proc/reader_linux.go]: Ok, but we are adding a mutex for something that will probably never be used concurrently... but not a big deal.

File: probe/proc/reader_linux.go (hunk 9 above)
Comment (inercia) [inline probe/proc/reader_linux.go]: Yes, it is a cache for `/proc/something`, but for file handles, not for content. The most common case is when we read from `/proc/<some_proc_id>/[comm, cmdline, stat]`: we try to get from the cache and hopefully we will get an already-open file handle. A `60 * time.Second` expiration time would mean that we could have entries in the cache for up to 60 seconds that will never be used...

Further improvements would be to cache the content for `comm` and `cmdline`, and we could even cache the `stat` info for `/proc/<some_proc_id>/fd/*`.

File: probe/proc/reader_darwin.go (hunk 10 above)
Comment (inercia) [inline probe/proc/reader_darwin.go]: Fixed.

File: probe/proc/reader_linux.go (hunk 9 above)
Comment (peterbourgon) [inline probe/proc/reader_linux.go]: Gotcha!

**Comment by peterbourgon** [comment]
:+1: pending Coveralls. (Maybe Coveralls is broken?)

**Comment by inercia** [comment]
@peterbourgon I think the coveralls hook has failed (the jobs is finisished, but the "Coverage pending" is still here). I'm merging....

**Pull request merged**